File: mac/Classes/Model/LRManifestBasedObject.m

#import "LRManifestBasedObject.h"
#import "Errors.h"


@implementation LRManifestBasedObject {
    NSMutableArray *_errors;
}

- (instancetype)initWithManifest:(NSDictionary *)manifest errorSink:(id<LRManifestErrorSink>)errorSink {
    self = [super init];
    if (self) {
        _errorSink = errorSink;
        _manifest = [manifest copy];
        _valid = YES;
    }
    return self;
}

- (void)addErrorMessage:(NSString *)message {
    message = [NSString stringWithFormat:@"%@ in %@ %@", message, NSStringFromClass(self.class), self.manifest];

    _valid = NO;
    if (!_errors) {
        _errors = [NSMutableArray new];
    }
    [_errors addObject:[NSError errorWithDomain:LRErrorDomain code:LRErrorInvalidManifest userInfo:@{NSLocalizedDescriptionKey:message}]];

    if (_errorSink) {
        [_errorSink addErrorMessage:message];
    }
}

- (NSArray *)errors {
    return [_errors copy];
}

@end

File: LiveReload/LRActionKit/Source/Basics/LRManifestBasedObject.swift
import Foundation

public class LRManifestBasedObject : NSObject, LRManifestErrorSink {

    public let manifest: [String: AnyObject]

    private weak var errorSink: LRManifestErrorSink?

    public var errors: [NSError] = []

    public init(manifest: [String: AnyObject], errorSink: LRManifestErrorSink?) {
        self.errorSink = errorSink
        self.manifest = manifest
        super.init()
    }

    public func addErrorMessage(message: String) {
        let fullMessage = "\(message) in \(self.dynamicType) \(manifest)"

        errors.append(NSError(domain: ActionKitErrorDomain, code: ActionKitErrorCode.InvalidManifest.rawValue, userInfo: [NSLocalizedDescriptionKey: fullMessage]))

        if let es = errorSink {
            es.addErrorMessage(message)
        }
    }

    public var valid: Bool {
        return errors.isEmpty
    }

}
